adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java

```diff
@@ -0,0 +1,167 @@
+                        return code2.compareTo(code1);
+                    }
+                }
+            });//TODO: comparator!
+            sortedArchetypeTermMap.putAll(stringArchetypeTermMap);
+            termDefinitions.put(language, sortedArchetypeTermMap);
+
+        }
+    }
+
+    private void walkArchetype(String sourceFile, Archetype archetype, Archetype resultingArchetype) {
+        Stack<CObject> workList = new Stack<>();
+        workList.push(archetype.getDefinition());
+        List<CObject> cObjects = new ArrayList<>();
+        while(!workList.empty()) {
+            CObject next = workList.pop();
+            if(!terminologyHasCode(archetype, next)) {
+                cObjects.add(next);
+                addCodeToTerminology(sourceFile, resultingArchetype, next);
+            }
+            for(CAttribute attribute:next.getAttributes()) {
+                for(CObject child:attribute.getChildren()) {
+                    workList.push(child);
+                }
+            }
+        }
+    }
+
+    private void addCodeToTerminology(String sourceFile, Archetype resultingArchetype, CObject next) {
+        String text = getCommentName(sourceFile, next.getNodeId());
+        if(text == null) {
+            text = "X";
+        }
+        String description = text; //TODO: grep comment from sourcefile!
+        Map<String, Map<String, ArchetypeTerm>> termDefinitions = resultingArchetype.getTerminology().getTermDefinitions();
+        for(String language: termDefinitions.keySet()) {
+            if(termDefinitions.get(language).get(next.getNodeId()) == null) {
+                ArchetypeTerm newCode = new ArchetypeTerm();
+                newCode.setCode(next.getNodeId());
+                newCode.setText(text);
+                newCode.setDescription(description);
+                termDefinitions.get(language).put(next.getNodeId(), newCode);
+            }
+        }
+    }
+
+    private String getCommentName(String sourceFile, String nodeId) {
+        String[] lines = sourceFile.split("\n");
+        for(String line:lines) {
+            Matcher matcher = commentPattern.matcher(line);
+            if(matcher.matches()) {
+                System.out.println("we have a match!");
+                String idcode = matcher.group("idcode");
+                String comment = matcher.group("comment");
+                if(idcode.equalsIgnoreCase(nodeId)) {
+                    return comment.trim();
+                }
+            }
+
+        }
+        return null;
+    }
+
+    public boolean terminologyHasCode(Archetype archetype, CObject cObject) {
+        String nodeId = cObject.getNodeId();
+        int codeSpecializationDepth = AOMUtils.getSpecializationDepthFromCode(nodeId);
+        int archetypeSpecializationDepth = archetype.specializationDepth();
+        if(codeSpecializationDepth > archetypeSpecializationDepth) {
+            return true;//not exactly, this is a validation failure that needs to be fixed. log?
```

Comment:
Can this be fixed now? Or should a new ticket be created for this?

Reply:
changed comment with explanation - this cannot be fixed, must remain as is and will be logged by the archetype validator.